### DISCUSSION TITLE
Query optimisation (CQL backend)

### DISCUSSION THREAD

**Discussion by sergei-riazanov-privitar** [JanusGraph/janusgraph#3000] [General]
Hi guys, I'm trying to optimise the following query.

I have a collection of secondary IDs (with index on it). For each ID, I need to:
1. Find the vertex

2. Find out edges

3. Get data from that side of the edges.

I'm doing it with this query:
```
g.V().has(ID, P.within(ids))
                         .where(outE(LABEL).
                         .project("parentObjectId", "tags")
                         .by(values(ID))
                         .by(outE(LABEL).inV().fold())
                         .toList()
```
However, this query is almost as slow as if I iterate all the vertexes and query the edges for each of them sequentially. What I'm doing wrong?

**Comment by vtslab** [comment]
Can you add the output of the profile() step?
You can also check older performance related questions on the user forum, e.g. 
[https://lists.lfaidata.foundation/g/janusgraph-users/search?p=%252C%252Cbatch%252C20%252C2%252C0%252C0&q=query.batch]https://lists.lfaidata.foundation/g/janusgraph-users/search?p=%252C%252Cbatch%252C20%252C2%252C0%252C0&q=query.batchl)
